packages/language/src/utils/search.ts

```diff
@@ -4,23 +4,57 @@ export function binaryTokenSearch(
   tokens: IToken[],
   offset: number,
 ): IToken | undefined {
+  return tokens[binaryTokenIndexSearch(tokens, offset)];
+}
+
+export function binaryTokenIndexSearch(
+  tokens: IToken[],
+  offset: number,
+): number {
   let low = 0;
   let high = tokens.length - 1;
+  let token: IToken | undefined;
+  let mid = -1;
   while (low <= high) {
-    const mid = Math.floor((low + high) / 2);
-    const token = tokens[mid];
+    mid = Math.floor((low + high) / 2);
+    token = tokens[mid];
     const start = token.startOffset;
     const end = token.endOffset!;
-    if (start <= offset && offset <= end) {
-      return token;
-    } else if (offset - end === 1 && /\w$/u.test(token.image)) {
-      // If the offset is right after the end of a word token, return that token
-      return token;
+    if ((start <= offset && offset <= end) || isAtTokenEnd(token, offset)) {
+      return mid;
     } else if (start > offset) {
       high = mid - 1;
     } else {
       low = mid + 1;
     }
   }
-  return undefined;
+  return -1;
+}
+
+function isAtTokenEnd(token: IToken, offset: number): boolean {
+  const end = token.endOffset!;
+  // If the offset is right after the end of a word token, return that token
+  return offset - end === 1 && /\w$/u.test(token.image);
+}
+
+function isBeforeTokenEnd(token: IToken, offset: number): boolean {
+  return token.endOffset! >= offset || isAtTokenEnd(token, offset);
+}
+
+export function completionTokenSearch(
```

Comment:
Can call this `completionTokenIndexSearch`, matching the naming done above for the modified binary token search.